server/subscribe.go

```diff
@@ -0,0 +1,472 @@
+		}
+
+		subs.Channel.remove(&subscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) addChannelSubscription(newSubscription *ChannelSubscription) error {
+	return p.atomicModify(JIRA_SUBSCRIPTIONS_KEY, func(initialBytes []byte) ([]byte, error) {
+		subs, err := SubscriptionsFromJson(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		newSubscription.Id = model.NewId()
+		subs.Channel.add(newSubscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) editChannelSubscription(modifiedSubscription *ChannelSubscription) error {
+	return p.atomicModify(JIRA_SUBSCRIPTIONS_KEY, func(initialBytes []byte) ([]byte, error) {
+		subs, err := SubscriptionsFromJson(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		oldSub, ok := subs.Channel.ById[modifiedSubscription.Id]
+		if !ok {
+			return nil, errors.New("Existing subscription does not exist.")
+		}
+		subs.Channel.remove(&oldSub)
+		subs.Channel.add(modifiedSubscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) atomicModify(key string, modify func(initialValue []byte) ([]byte, error)) error {
+	readModify := func() ([]byte, []byte, error) {
+		initialBytes, appErr := p.API.KVGet(key)
+		if appErr != nil {
+			return nil, nil, errors.Wrap(appErr, "unable to read inital value")
+		}
+
+		modifiedBytes, err := modify(initialBytes)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "modification error")
+		}
+
+		return initialBytes, modifiedBytes, nil
+	}
+
+	success := false
+	for !success {
+		//initialBytes, newValue, err := readModify()
+		_, newValue, err := readModify()
+		if err != nil {
+			return err
+		}
+
+		var setError *model.AppError
+		// Commenting this out so we can support < 5.12 for 2.0
+		//success, setError = p.API.KVCompareAndSet(key, initialBytes, newValue)
+		setError = p.API.KVSet(key, newValue)
+		success = true
+		if setError != nil {
+			return errors.Wrap(setError, "problem writing value")
+		}
+
+	}
+
+	return nil
+}
+
+func httpSubscribeWebhook(p *Plugin, w http.ResponseWriter, r *http.Request) (int, error) {
+	if r.Method != http.MethodPost {
+		return http.StatusMethodNotAllowed,
+			fmt.Errorf("Request: " + r.Method + " is not allowed, must be POST")
+	}
+
+	cfg := p.getConfig()
+	if cfg.Secret == "" || cfg.UserName == "" {
+		return http.StatusForbidden, fmt.Errorf("JIRA plugin not configured correctly; must provide Secret and UserName")
+	}
+
+	if subtle.ConstantTimeCompare([]byte(r.URL.Query().Get("secret")), []byte(cfg.Secret)) != 1 {
+		return http.StatusForbidden, fmt.Errorf("Request URL: secret did not match")
+	}
+
+	defer r.Body.Close()
```

Comment:
```
// For server requests, the Request Body is always non-nil
// but will return EOF immediately when no body is present.
// The Server will close the request body. The ServeHTTP
// Handler does not need to.
```

Reply:
That was added at my request, before learning that it's not needed server-side.